### PR TITLE
Features/prune full2

### PR DESCRIPTION
**Pull request by liuhaox1**

### CODE REVIEW

File: cmd/gzv/cli/miner_rpc.go [hunk 1]
```diff
@@ -72,13 +72,14 @@ func startHTTP(endpoint string, apis []rpc.API, modules []string, cors []string,
 	if endpoint == "" {
 		return nil
 	}
+	isPruneMode := core.BlockChainImpl.IsPruneMode()
```
Comment (bird512) [inline cmd/gzv/cli/miner_rpc.go:75]: we can get the prune mode config in current package directly. do not need to add this dependence for core package

File: core/blockchain.go [hunk 2]
```diff
@@ -68,19 +70,33 @@ type BlockChainConfig struct {
...
+	blocks          *tasdb.PrefixedDatabase
+	blockHeight     *tasdb.PrefixedDatabase
+	txDb            *tasdb.PrefixedDatabase
+	stateDb         *tasdb.PrefixedDatabase
+	smallStateDb    *smallStateStore
+	cacheDb         *tasdb.PrefixedDatabase
+	batch           tasdb.Batch
+	triegc          *prque.Prque // Priority queue mapping block numbers to tries to gc
+	stateCache      account.AccountDatabase
+	running         int32 // running must be called atomically
```
Comment (bird512) [inline core/blockchain.go]: better rename to shutdowning

File: core/trie_small_db_store.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+		return fmt.Errorf("store trie pure copy info error %v", err)
+	}
+	return nil
+}
+
+// HasStateData check the small db exists data
+func (store *smallStateStore) HasStateData() bool {
+	iter := store.db.NewIterator()
+	defer iter.Release()
+	return iter.Seek([]byte(smallDbRootDatas))
```
Comment (bird512) [inline core/trie_small_db_store.go]: should check the result's prefix after Seek. Seek may return data  of key greater than or equal to the given key.

File: core/blockchain.go [hunk 4]
```diff
@@ -250,6 +308,7 @@ func initBlockChain(helper types.ConsensusHelper, minerAccount types.Account) er
 	BlockChainImpl = chain
 
 	// db cache enabled
+	iteratorNodeCacheSize := 30000
```
Comment (bird512) [inline core/blockchain.go:336]: below if is always true

File: core/blockchain_action.go [hunk 5]
```diff
@@ -262,14 +263,176 @@ func (chain *FullBlockChain) validateBlock(source string, b *types.Block) (bool,
 	return true, nil
 }
 
+// DeleteSmallDbDataByRoots will delete root list if we reset tops
+func (chain *FullBlockChain) DeleteSmallDbDataByRoots(roots []common.Hash) {
+	if len(roots) == 0 {
```
Comment (bird512) [inline core/blockchain_action.go]: this if check can be removed

File: core/blockchain.go [hunk 6]
```diff
@@ -418,10 +471,63 @@ func (chain *FullBlockChain) GetConsensusHelper() types.ConsensusHelper {
...
+		}
+		preHash := bh.PreHash
+		bh = chain.queryBlockHeaderByHash(preHash)
+		if bh == nil {
+			return nil, fmt.Errorf("find block hash not exists,block hash is %v", preHash)
+		}
+		if bh.Height == 0 {
+			return bh, nil
+		}
+		_, e = chain.accountDBAt(bh.Height)
```
Comment (sesamegu) [inline core/blockchain.go]: _, e = chain.accountDBAt(bh.Height)
this is not need

File: core/trie_small_db_store.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+import (
+	"bytes"
+	"fmt"
+	"github.com/zvchain/zvchain/common"
+	"github.com/zvchain/zvchain/storage/tasdb"
+	"sync"
+)
+
+var (
+	persistentHeight = "ph"
```
Comment (bird512) [inline core/trie_small_db_store.go]: Please add some comments for these three prefixes

File: storage/trie/database.go [hunk 8]
```diff
@@ -316,38 +408,99 @@ func (db *NodeDatabase) insertPreimage(hash common.Hash, preimage []byte) {
 	db.preimagesSize += common.StorageSize(common.HashLength + len(preimage))
 }
 
+func (db *NodeDatabase) tryGetFromCache(hash common.Hash) []byte {
+	// reset the meter if read count > 2000000, for more accuracy statistic about recent db read
```
Comment (bird512) [inline storage/trie/database.go]: This comment is not match to below code

File: storage/trie/database.go [hunk 9]
```diff
@@ -701,12 +932,13 @@ func (db *NodeDatabase) Size() (common.StorageSize, common.StorageSize) {
...
-	var flushlistSize = common.StorageSize((len(db.nodes) - 1) * 2 * common.HashLength)
-	return db.nodesSize + flushlistSize, db.preimagesSize
+	// counted.
+	var metadataSize = common.StorageSize((len(db.nodes) - 1) * cachedNodeSize)
+	var metarootRefs = common.StorageSize(len(db.nodes[common.Hash{}].children) * (common.HashLength + 2))
+	return db.nodesSize + db.childrenSize + metadataSize - metarootRefs, db.preimagesSize
 }
 
-// verifyIntegrity is a debug method to iterate over the entire trie stored in
+// traverse is a debug method to iterate over the entire trie stored in
```
Comment (bird512) [inline storage/trie/database.go:943]: comments  traverse not match the method name verifyIntegrity

File: storage/trie/database.go [hunk 10]
```diff
@@ -747,3 +979,34 @@ func (db *NodeDatabase) accumulate(hash common.Hash, reachable map[common.Hash]s
 		db.accumulate(child, reachable)
 	}
 }
+
+func (db *NodeDatabase) printPath(root common.Hash, height uint64, rets []common.Hash) {
```
Comment (bird512) [inline storage/trie/database.go:985]: This function is never used, should add some comments for it

File: core/blockchain.go [hunk 11]
```diff
@@ -418,10 +499,62 @@ func (chain *FullBlockChain) GetConsensusHelper() types.ConsensusHelper {
...
+			}
+			cnt++
+		}
+		preHash := bh.PreHash
+		bh = chain.queryBlockHeaderByHash(preHash)
+		if bh == nil {
+			return nil, fmt.Errorf("find block hash not exists,block hash is %v", preHash)
+		}
+		if bh.Height == 0 {
+			return bh, nil
```
Comment (bird512) [inline core/blockchain.go]: should check the value of cnt before return directly

File: storage/trie/database.go [hunk 12]
```diff
@@ -294,7 +392,9 @@ func (db *NodeDatabase) insert(hash common.Hash, blob []byte, node node) {
 		}
 	}
 	db.nodes[hash] = entry
-
+	if db.enablePrune {
+		db.commitFullNodes = append(db.commitFullNodes, &storeBlob{Key: hash, Raw: entry.rlp()})
```
Comment (bird512) [inline storage/trie/database.go]: do not need to rlp it again. can use copy()

File: core/blockchain_action.go [hunk 13]
```diff
@@ -262,14 +263,169 @@ func (chain *FullBlockChain) validateBlock(source string, b *types.Block) (bool,
...
+// mergeSmallDbDataToBigDB is for cold start
+// Begin is last persistent height,end is top height,between two heights block state data from small db to big db
+func (chain *FullBlockChain) mergeSmallDbDataToBigDB(top *types.BlockHeader) error {
+	// get the last persistent height from small db
+	lastStateHeight := chain.smallStateDb.GetStatePersistentHeight()
+	if top == nil && lastStateHeight == 0{
+		return nil
+	// if big db is deleted,but small db not be deleted,we not support this stage!
+	}else if top == nil && lastStateHeight > 0{
+		return  fmt.Errorf("db is damaged,suggest delete d_mall and try again")
```
Comment (bird512) [inline core/blockchain_action.go]: Should print this error message to console

**Review by bird512 — DISMISSED** [review]

**Review by darren0718 — APPROVED** [review]

**Review by bird512 — APPROVED** [review]

**Pull request merged**